docs/slices.md

```diff
@@ -0,0 +1,1881 @@
+a[:]
+a[x::]
+a[x::z]
+```
+
+Furthermore, for a slice `a[x:y:z]`, if `a` is a Python built-in object or a
+NumPy array, there is an additional semantic restriction, which is that the
+expressions `x`, `y`, and `z` must be integers.
+
+The three arguments to a slice are traditionally called `start`, `stop`, and
+`step`:
+
+```py
+a[start:stop:step]
+```
+
+We will use these names throughout this guide.
+
+It is worth noting that the `x:y:z` syntax is not valid outside of square
+brackets. However, slice objects can be created manually using the `slice()`
+builtin (`a[x:y:z]` is the same as `a[slice(x, y, z)]`). You can also use the
+[`ndindex.Slice()`](slice-api) object if you want to perform more advanced
+operations.
+
+(rules)=
+## Rules
+
+These are the rules to keep in mind to understand how slices work. Each of
+these is explained in detail below. Many of the detailed descriptions below
+also outline several *wrong* rules, which are bad ways of thinking about
+slices but which you may be tempted to think about as rules. The below 7 rules
+are always correct.
+
+In this document, "*nonnegative*" means $\geq 0$ and "*negative*" means $< 0$.
+
+For a slice `a[start:stop:step]`:
+
+1. `start` and `step` use **0-based indexing** from the **start** of `a`
+   when they are **nonnegative**, and **−1-based indexing** from **end** of
+   `a` when they are **negative**. (See sections {ref}`0-based` and
+   {ref}`negative-indices`)
+2. `stop` is never included in the slice. (See section {ref}`half-open`)
+3. `start` and `stop` are clipped to the bounds of `a`. (See section {ref}`clipping`)
+4. The slice starts at `start` and successively adds `step` until it reaches
+   an index that is at or past `stop`, and then stops without including that
+   `stop` index. (See sections {ref}`steps` and {ref}`negative-steps`)
+5. If `step` is omitted it defaults to `1`. (See section {ref}`omitted`)
+6. If `start` or `stop` are omitted they extend to the beginning or end of `a`
+   in the direction being sliced. Slices like `a[:i]` or `a[i:]` should be
+   though of as the `start` or `stop` being omitted, not as a colon to the
+   left or right of an index. (See section {ref}`omitted`)
+7. Slicing something never raises an `IndexError`, even if the slice is empty.
+   For a NumPy array, a slice always keeps the axis being sliced, even if that
+   means the resulting dimension will be 0 or 1. (See section {ref}`subarray`)
+
+(integer-indices)=
+## Integer indices
+
+To understand slices, it is good to first review how integer indices work.
+Throughout this guide, we will use as an example this prototype list:
+
+$$
+a = [\mathtt{\textsf{'}a\textsf{'}},\ \mathtt{\textsf{'}b\textsf{'}},\ \mathtt{\textsf{'}c\textsf{'}},\
+\mathtt{\textsf{'}d\textsf{'}},\ \mathtt{\textsf{'}e\textsf{'}},\ \mathtt{\textsf{'}f\textsf{'}},\ \mathtt{\textsf{'}g\textsf{'}}]
+$$
+
+The list `a` has 7 elements.
+
+The elements of `a` are strings, but the indices and slices on `a` will always
```

Comment:
"on `a` will" looks awkward in the rendered form, because the inline code font is not distinct enough. I suggest "on the list will"

Reply:
I can fix this, but we should also consider if we can change the code font to be more distinct.

Reply:
I also tried to avoid saying "list" too much because the rules here apply to lists, strings, and arrays.

Reply:
yeah, maybe "sequence" instead of list, but that doesn't really have the same tone